### PR TITLE
[WB-8246,WB-8572] Support version IDs in artifact refs, fix s3/gcs references in Windows

### PR DESCRIPTION
**Pull request by annirudh**
https://wandb.atlassian.net/browse/WB-8246
https://wandb.atlassian.net/browse/WB-8572

Description
-----------
Allows users to add a specific version of an object if bucket versioning is enabled:
- `s3://bucket/key?versionID=...` ([docs](https://docs.aws.amazon.com/AmazonS3/latest/userguide/RetrievingObjectVersions.html))
- `gs://bucket/key#generation` ([docs](https://cloud.google.com/storage/docs/using-versioned-objects#prereq-rest))

Fixes a bug with artifact references in Windows where they are populated with `os.path.join`, which uses the incorrect separator for S3/GCS references.

Testing
-------
- [x] New unit tests
- [x] Tested adding a version in S3 + GCS

### CODE REVIEW

File: wandb/sdk/wandb_artifacts.py [hunk 1]
```diff
@@ -1544,11 +1584,12 @@ def init_gcs(self) -> "gcs_module.client.Client":
         self._client = storage.Client()
         return self._client
 
-    def _parse_uri(self, uri: str) -> Tuple[str, str]:
+    def _parse_uri(self, uri: str) -> Tuple[str, str, Optional[str]]:
         url = urlparse(uri)
         bucket = url.netloc
         key = url.path[1:]
-        return bucket, key
+        version = url.fragment if url.fragment else None
```
Comment (kptkin) [inline wandb/sdk/wandb_artifacts.py:1591]: what is the falsey value of `url.fragment`? i.e. do you need the if/else?

File: wandb/sdk/wandb_artifacts.py [hunk 2]
```diff
@@ -1465,22 +1486,41 @@ def _entry_from_obj(
...
+        bucket, key, _ = self._parse_uri(path)
+
+        # Always use posix paths, since that's what S3 uses.
+        posix_key = pathlib.PurePosixPath(obj.key)  # the bucket key
+        posix_path = pathlib.PurePosixPath(bucket) / pathlib.PurePosixPath(
+            key
+        )  # the path, with the scheme stripped
+        posix_prefix = pathlib.PurePosixPath(prefix)  # the prefix, if adding a prefix
+        posix_name = pathlib.PurePosixPath(name or "")
+        posix_ref = posix_path
```
Comment (kptkin) [inline wandb/sdk/wandb_artifacts.py:1506]: nit: should this one be in the else condition?

File: wandb/sdk/wandb_artifacts.py [hunk 3]
```diff
@@ -1608,13 +1649,23 @@ def store_path(
...
+            raise ValueError(
+                f"Specifying a versionId is not valid for s3://{bucket} as it does not have versioning enabled."
+            )
 
         if not checksum:
             return [ArtifactManifestEntry(name or key, path, digest=path)]
+
         start_time = None
-        obj = self._client.bucket(bucket).get_blob(key)
+        obj = self._client.bucket(bucket).get_blob(key, generation=version)
```
Comment (kptkin) [inline wandb/sdk/wandb_artifacts.py:1668]: what happens if the `version` is`None`?

**Review by vanpelt — APPROVED** [review]
🔥

File: wandb/sdk/wandb_artifacts.py (hunk 1 above)
Comment (annirudh) [reply wandb/sdk/wandb_artifacts.py:1591]: The falsey value is `''`, but I need it to be `None` hence this ugly ternary op.

File: wandb/sdk/wandb_artifacts.py (hunk 2 above)
Comment (annirudh) [reply wandb/sdk/wandb_artifacts.py:1506]: Ah you mean explicitly add this as part of the `else` right after the `elif multi` case? That might be cleaner.

**Pull request merged**